internal/blueprint/repository_customizations.go

```diff
@@ -326,3 +330,84 @@ func validateCustomRepository(repo *RepositoryCustomization) error {
 	}
 	return nil
 }
+
+// DeepCopy returns a deep copy of the given slice of RepositoryCustomization
+// Since we are replacing the gpgkeys of the customization with the file path
+// we don't want to modify the original slice.
+func DeepCopyCustomRepositories(repos []RepositoryCustomization) []RepositoryCustomization {
+	clonedRepos := make([]RepositoryCustomization, len(repos))
+	copy(clonedRepos, repos)
+	for idx, repo := range clonedRepos {
+		// copy the baseurls
+		baseurls := make([]string, len(repo.BaseURLs))
+		copy(baseurls, repo.BaseURLs)
+		clonedRepos[idx].BaseURLs = baseurls
+
+		// copy the gpgkeys
+		gpgkeys := make([]string, len(repo.GPGKeys))
+		copy(gpgkeys, repo.GPGKeys)
+		clonedRepos[idx].GPGKeys = gpgkeys
+	}
+	return clonedRepos
+}
+
+func GetInlineGPGKeysFromCustomRepositories(repos []RepositoryCustomization) ([]*fsnode.File, error) {
+	files := []*fsnode.File{}
+	for _, repo := range repos {
+		for keyIdx, gpgkey := range repo.GPGKeys {
+			// check if the gpgkey is an inline file
+			if _, ok := url.ParseRequestURI(gpgkey); ok != nil {
+				// create the file path
+				path := fmt.Sprintf("/etc/pki/rpm-gpg/RPM-GPG-KEY-%s-%d", repo.Id, keyIdx)
+				// replace the gpgkey with the file path
+				repo.GPGKeys[keyIdx] = fmt.Sprintf("file://%s", path)
+				// create the fsnode for the gpgkey file
+				file, err := fsnode.NewFile(path, common.ToPtr(os.FileMode(0777)), "root", "root", []byte(gpgkey))
+				if err != nil {
+					return nil, err
+				}
+				files = append(files, file)
+			}
+		}
+	}
+	return files, nil
+}
+
+func RepoCustomizationsToRepoConfig(repos []RepositoryCustomization) map[string][]rpmmd.RepoConfig {
+	if len(repos) == 0 {
+		return nil
+	}
+
+	m := make(map[string][]rpmmd.RepoConfig)
+
+	for _, repo := range repos {
+		convertedRepo := repo.customRepoToRepoConfig()
+		m[repo.Filename] = append(m[repo.Filename], convertedRepo)
```

Comment:
If I read the implementation correctly, `repo.Filename` may be an empty string. It could make sense to add a method for the `RepositoryCustomization` struct which will generate one if not set explicitly by user. We discussed that we could default to the `<repo_id>.repo`.

Reply:
So at the moment, I'm setting the filename in the unmarshal functions if nothing is set - it might not the best place to do it and I could make that more explicit, like you suggest.

For reference:
https://github.com/osbuild/osbuild-composer/blob/fd61e02f2a3a86117a9147a2dd59040d06a3ed60/internal/blueprint/repository_customizations.go#L195

Reply:
Ah, I indeed missed that. I think that doing this in the custom unmarshal function may not be the best place. Specifically, because this will solve the problem of filling the not provided filename in CloudAPI where we just copy customizations from the compose request to the blueprint created on the fly https://github.com/osbuild/osbuild-composer/blob/03b093acb024675eab75b7fe266b3e6db2504dc7/internal/cloudapi/v2/handler.go#L142

We do not unmarshal there into the BP customizations so the missing filename would not be filled in in such case.

WRT the filename if it is not provided, I'll let others to comment as well, but I find a bit nicer if we simply use the repo ID as the repo filename.

In addition, the repo IDs must be unique across all configuration files, which is not currently enforced in any way and it would probably make sense to check as it would solve some corner cases. To complicate things more, the user-provided repo ids can conflict with repos specified in the image configuration (currently used only by GCE images).

Reply:
Yeah no problem at all... I like your solution better, it's simpler and more elegant and avoids magic. I've fixed that already and will push the changes in a bit :). 

It's also less of headache if we just save them to a repo with their own id rather than grouping all the repos with the same filename together.